src/device_scan.cc

```diff
@@ -27,25 +27,87 @@
 */
 
 #include "fty_discovery_classes.h"
-void device_scan_scan (CIDRList *listAddr, zconfig_t *config, discovered_devices_t *devices, zsock_t *pipe)
+#include <algorithm>
+bool device_scan_scan (zlist_t *listScans, discovered_devices_t *devices, zsock_t *pipe)
 {
-    zlist_t *args = zlist_new();
-    zlist_append(args, listAddr);
-    zlist_append(args, config);
-    zlist_append(args, devices);
-    zactor_t *scan_nut = zactor_new (scan_nut_actor, args);
-    zpoller_t *poller = zpoller_new (pipe, scan_nut, NULL);
+    CIDRList *listAddr = (CIDRList *) zlist_first(listScans);
+    zpoller_t *poller = zpoller_new(pipe, NULL);
+    std::vector<zactor_t *> listActor;
+    bool term = false;
+    while(listAddr != NULL) {
+        zlist_t *args = zlist_new();
+        zlist_append(args, listAddr);
+        zlist_append(args, devices);
+        zactor_t *scan_nut = zactor_new (scan_nut_actor, args);
+        zpoller_add(poller, scan_nut);
+        listActor.push_back(scan_nut);
+        listAddr = (CIDRList *) zlist_next(listScans);
+    }
+
+    size_t number_end_actor = 0;
+
     while (!zsys_interrupted) {
         void *which = zpoller_wait (poller, 5000);
-        if(which == scan_nut) {
-            zmsg_t *msg = zmsg_recv (scan_nut);
+        if (which == pipe) {
+            zmsg_t *msg = zmsg_recv (pipe);
+            if(msg) {
+                char *cmd = zmsg_popstr (msg);
+                if(cmd) {
+                    if(streq (cmd, "$TERM")) {
+                        zstr_free(&cmd);
+                        zmsg_destroy (&msg);
+                        term = true;
+                        break;
+                    }
+                    zstr_free(&cmd);
+                }
+                zmsg_destroy (&msg);
+            }
+        } else if (which == NULL) {
+            //time out, nothing to do
+        }else { //any of scan_nut_actor
+            zmsg_t *msg = zmsg_recv (which);
             if (msg) {
                 char *cmd = zmsg_popstr (msg);
                 if (cmd) {
-                    if (streq (cmd, "$TERM") || streq (cmd, "DONE")) {
+                    if (streq (cmd, "$TERM")) {
                         zstr_free (&cmd);
                         zmsg_destroy (&msg);
                         break;
+                    } else if (streq (cmd, "DONE")) {
```

Comment:
All actions can be defined in #define??? with a comment, avoid some mistakes ? more over these are used in mutiple files.